docs/proposals/container-runtime-interface-v1.md

```diff
@@ -0,0 +1,215 @@
+
+
+Below is an example of the proposed interfaces.
+```go
+// PodSandboxManager contains basic operations for sandbox.
+type PodSandboxManager interface {
+	Create(config *PodSandboxConfig) (string, error)
+	Delete(id string) (string, error)
+	List(filter PodSandboxFilter) []PodSandboxListItem
+	Status(id string) PodSandboxStatus
+}
+
+// ContainerRuntime contains basic operations for containers.
+type ContainerRuntime interface {
+    Create(config *ContainerConfig, sandboxConfig *PodSandboxConfig, sandboxIDng) (string, error)
+    Start(id string) error
+    Stop(id string, timeout int) error
+    Remove(id string) error
+    List(filter ContainerFilter) ([]ContainerListItem, error)
+    Status(id string) (ContainerStatus, error)
+}
+
+// ContainerCommandExecutor provides methods to run commands in a container.
+type ContainerCommandExecutor interface {
+	Exec(id string, cmd []string, streamOpts StreamOptions) error
+}
+
+// ImageOperations contains image-related operations.
+type ImageOperations interface {
+	List() ([]Image, error)
+	Pull(image ImageSpec, auth AuthConfig) error
+	Remove(image ImageSpec) error
+	Status(image ImageSpec) (Image, error)
+    Metrics(image ImageSpec) (ImageMetrics, error)
+}
+
+type ContainerMetricsGetter interface {
+    ContainerMetrics(id string) (ContainerMetrics, error)
+}
+
+```
+
+### Pod Lifecycle
+
+The PodSandbox’s lifecycle is decoupled from the containers, i.e., a sandbox
+is created before any containers, and can exist after all containers in it have
+terminated.
+
+Assume there is a pod with a single container C. To start a pod:
+```
+  create sandbox Foo --> create container C --> start container C
+```
+
+To delete a pod:
+```
+  stop container C --> remove container C --> delete sandbox Foo
+```
+
+Kubelet is responsible for creating, starting the containers based on the
+pod-level restart policy. E.g., if container C dies unexpectedly, Kubelet
+will simply create and start the container again in sandbox Foo.
+
+Kubelet is also responsible for gracefully terminating all the containers
+in the sandbox before deleting the sandbox. If Kubelet chooses to delete
+the sandbox with running containers in it, those containers should be forcibly
+deleted.
+
+Note that every PodSandbox/container lifecycle operation (create, start,
+stop, delete) should either returns an error or blocks until the operation
+succeeds. A successful operation should include a state transition of the
+PodSandbox/container. E.g., if a `Create` call for a container does not
+return an error, the container state should be "created" when the runtime is
+queried.
+
+### Updates to PodSandbox or Containers
+
+Kubernetes support updates only to a very limited set of fields in the Pod
+Spec.  These updates may require containers to be re-created by Kubelet. This
+can be achieved through the proposed, imperative container-level interface.
+On the other hand, PodSandbox update currently is not required.
+
+
+### Container Lifecycle Hooks
```

Comment:
The hope is that we can run volume and network hooks in pre-start and post-stop phase. Will that work? What are the explicit requirements for these plugins other than mount and network namespaces being present?

> Moreover, can we have a hook for periodic sync? That would allow plugins to assert the state periodically instead of waiting for event trigger.

Shouldn't this be built on top this interface? Invoke exec periodically, akin to existing probes.